### ISSUE TITLE
[rrd4j] Custom datasource setups seem not to be used.

### ISSUE THREAD

**Issue by JueBag** [openhab/openhab-addons#8771]
In order to better understand the new rrd4j version I did some test trials with custom datasource setups (especially archive setups).
It seems to me as if those setups are read by openhab since they aare creating log lines like:
`2020-10-16 19:05:16.778 [DEBUG] [d4j.internal.RRD4jPersistenceService] - Created logger1 = GAUGE heartbeat = 90 min/max = 0.0/NaN step = 60 3 archives(s) = [ MAX xff = 0.5 steps = 1 rows = 120 MAX xff = 0.5 steps = 10 rows = 24 MAX xff = 0.5 steps = 60 rows = 24] 1 items(s) = [Test3  ]`
Checking with the API-Explorer revealed however that values in archive 1 were persited every 5 seconds, that sounds as if the `default_other` datasource setup is used ( although the log-line for this one doens't show any item using this one).

Used rrd4j.cfg
```
Logger1.def=GAUGE,90,0,U,60
Logger1.archives=MAX,.5,1,120:MAX,.5,10,24:MAX,.5,60,24
Logger1.items=Test3
```
I would expect that using this setup during the first 120 minutes one value for every minute would be saved.
Additionally I would expect each datasource setup to log all items which are using this setup in the "Created.." -line (as posted above).   

Using openhab 3.0.0 #1964 on a Raspi3B

My overall intention is to help in writing the updated documentation for rrd4j (see [https://github.com/JueBag/openhab-addons/tree/main/bundles/org.openhab.persistence.rrd4j](here) )

**Comment by JueBag** [comment]
Further testing revealed (to me) that only the datasource `default_other` acts as a default. The two other declared datasources (default_numeric and default_quantifiable)  can be selected manually by creating an .items line in the rrd4j.cfg. (this was most most probable the case as well before the code change, however unnoticed at least be me).

**Comment by kaikreuzer** [comment]
Hey @JueBag,

I have to admit that I didn't yet use custom definitions with OH3 and instead tried to do sensible defaults that should suit most situations.
Just some comments for clarification:

> Checking with the API-Explorer revealed however that values in archive 1 were persited every 5 second

You should not confuse the persistence strategies with the rrd definitions - even if you do not have many slots in the rrd files, you can potentially persist values much more frequently. You just won't be able to retrieve it with that granularity again.

> Additionally I would expect each datasource setup to log all items which are using this setup in the "Created.." -line

This is happening already or am I missing something?

> Further testing revealed (to me) that only the datasource default_other acts as a default.

Depending on the item type, it [should choose one of the three definitions](https://github.com/openhab/openhab-addons/blob/main/bundles/org.openhab.persistence.rrd4j/src/main/java/org/openhab/persistence/rrd4j/internal/RRD4jPersistenceService.java#L296-L308) as the default. Are you certain that this is not working?

> A service folder was not existing

It should be there (but empty):

<img width="261" alt="Screenshot 2020-10-19 at 23 42 49" src="https://user-images.githubusercontent.com/3244965/96515002-cf5de980-1264-11eb-8f9d-7b77017e44b7.png">

> No clou were to find a log about what is going on.

The log is printed out on the console in Eclipse. You can enable persistence debug logging by adding 
```
  <logger name="org.openhab.persistence" level="DEBUG" />
```
to the logback.xml in the demo app project.

**Comment by JueBag** [comment]
Thanks for the feedback!

> You should not confuse the persistence strategies with the rrd definitions....

That is obvious, however the 5 second intervall seen in the API-Explorer made me believe that the default_other datasource has been used instead of the custom datasource (which would have used 60 seconds).

 
> > Additionally I would expect each datasource setup to log all items which are using this setup in the "Created.." -line
> 
> This is happening already or am I missing something?

Sort off, as I revealed in my second post items assigned by an .items line to a datasource are printed. An item that isn't mentioned in any .items line doesn't show that way. In my test environment (not in the IDE) the item that is supposed to use my custom dadasource shows in the the respective "Created.." line, however it seems to use the default_other datasource.

 
> Depending on the item type, it [should choose one of the three definitions](https://github.com/openhab/openhab-addons/blob/main/bundles/org.openhab.persistence.rrd4j/src/main/java/org/openhab/persistence/rrd4j/internal/RRD4jPersistenceService.java#L296-L308) as the default. Are you certain that this is not working?

On my testing environment it does not work that way, in the IDE I see it different, so NO I am not certain!
I'm trying to get some additional log messages at the specific point in there, however my personal problems with getting log prevented it. You gave me the needed pointer! I'll continue the tests.

> > A service folder was not existing
> It should be there (but empty)

As I haven't had that folder but made it from within Eclipse with no effect. Without that folder and especially the rrd4j.cfg really used I can't complete my tests. 
I guess I'll recreate from scratch! 

Will report back!

**Comment by JueBag** [comment]
1.

 
Missing `services` folder:
Manually creating it did work. IMHO this folder is missing in [openhab-distro](https://github.com/openhab/openhab-distro/tree/master/launch/app/runtime/conf), but I'm not sure if it should be created automatically from elsewhere.

2. 

In order to run openHAB with rrd4j from the IDE I added it as a dependency to the pom.xml  and added `org.openhab.persistence.rrd4j` as a run requirement to the app.bndrun.
In order to use breakpoints in the rrd4j-code I followed the [Debugging documetation](https://www.openhab.org/docs/developer/ide/eclipse.html#modifying-and-debugging-add-ons).
Starting the degug-session I get the following error.
`Status ERROR: bndtools.core code=0 org.openhab.persistence.rrd4j;version=[3.0.0,3.0.1) Not found in [bndtools.m2e.MavenWorkspaceRepository@4bedd2d8, bndtools.m2e.MavenImplicitProjectRepository@fc568c2, bnd-cache [D:/Development/openhab-master/git/openhab-distro/launch/app/target/tmp/LAUNCH/app/cnf/cache/5.2.0/bnd-cache r/w=false]] null`
What am I missing?

**Comment by kaikreuzer** [comment]
Ok, I have spent this evening on rrd4j and came up with quite some fixes and improvements in https://github.com/openhab/openhab-addons/pull/8815.

With these changes in place, I can see custom definitions being used as well as the three default datasources to be assigned correctly. Here's an excerpt from my trace log:
```
2:11:45.310 [main] TRACE o.o.p.r.i.RRD4jPersistenceService:464 - Processing config: def = COUNTER,900,0,U,60
22:11:45.310 [main] TRACE o.o.p.r.i.RRD4jPersistenceService:464 - Processing config: items = DemoNumber
22:11:45.311 [main] TRACE o.o.p.r.i.RRD4jPersistenceService:464 - Processing config: archives = AVERAGE,0.5,1,480:AVERAGE,0.5,10,144
22:11:45.311 [main] DEBUG o.o.p.r.i.RRD4jPersistenceService:492 - Created ctr24h = COUNTER heartbeat = 900 min/max = 0.0/NaN step = 60 2 archives(s) = [ AVERAGE xff = 0.5 steps = 1 rows = 480 AVERAGE xff = 0.5 steps = 10 rows = 144] 1 items(s) = [DemoNumber ]
22:11:45.311 [main] DEBUG o.o.p.r.i.RRD4jPersistenceService:492 - Created default_quantifiable = GAUGE heartbeat = 600 min/max = NaN/NaN step = 10 5 archives(s) = [ AVERAGE xff = 0.5 steps = 1 rows = 360 AVERAGE xff = 0.5 steps = 6 rows = 10080 AVERAGE xff = 0.5 steps = 90 rows = 36500 AVERAGE xff = 0.5 steps = 360 rows = 43800 AVERAGE xff = 0.5 steps = 8640 rows = 3650] 0 items(s) = []
22:11:45.312 [main] DEBUG o.o.p.r.i.RRD4jPersistenceService:492 - Created default_other = GAUGE heartbeat = 3600 min/max = NaN/NaN step = 5 4 archives(s) = [ LAST xff = 0.5 steps = 1 rows = 720 LAST xff = 0.5 steps = 12 rows = 10080 LAST xff = 0.5 steps = 180 rows = 35040 LAST xff = 0.5 steps = 2880 rows = 21900] 0 items(s) = []
22:11:45.312 [main] DEBUG o.o.p.r.i.RRD4jPersistenceService:492 - Created default_numeric = GAUGE heartbeat = 600 min/max = NaN/NaN step = 10 5 archives(s) = [ LAST xff = 0.5 steps = 1 rows = 360 LAST xff = 0.5 steps = 6 rows = 10080 LAST xff = 0.5 steps = 90 rows = 36500 LAST xff = 0.5 steps = 360 rows = 43800 LAST xff = 0.5 steps = 8640 rows = 3650] 0 items(s) = []
...
22:11:45.324 [safeCall-1] TRACE o.o.p.r.i.RRD4jPersistenceService:331 - Using rrd definition 'default_other = GAUGE heartbeat = 3600 min/max = NaN/NaN step = 5 4 archives(s) = [ LAST xff = 0.5 steps = 1 rows = 720 LAST xff = 0.5 steps = 12 rows = 10080 LAST xff = 0.5 steps = 180 rows = 35040 LAST xff = 0.5 steps = 2880 rows = 21900] 0 items(s) = []' for item 'DemoSwitch'.
22:11:45.351 [safeCall-1] TRACE o.o.p.r.i.RRD4jPersistenceService:331 - Using rrd definition 'default_other = GAUGE heartbeat = 3600 min/max = NaN/NaN step = 5 4 archives(s) = [ LAST xff = 0.5 steps = 1 rows = 720 LAST xff = 0.5 steps = 12 rows = 10080 LAST xff = 0.5 steps = 180 rows = 35040 LAST xff = 0.5 steps = 2880 rows = 21900] 0 items(s) = []' for item 'MagicOnOff'.
22:11:45.359 [safeCall-1] TRACE o.o.p.r.i.RRD4jPersistenceService:331 - Using rrd definition 'default_other = GAUGE heartbeat = 3600 min/max = NaN/NaN step = 5 4 archives(s) = [ LAST xff = 0.5 steps = 1 rows = 720 LAST xff = 0.5 steps = 12 rows = 10080 LAST xff = 0.5 steps = 180 rows = 35040 LAST xff = 0.5 steps = 2880 rows = 21900] 0 items(s) = []' for item 'DemoDimmer'.
22:11:45.365 [safeCall-1] TRACE o.o.p.r.i.RRD4jPersistenceService:331 - Using rrd definition 'default_other = GAUGE heartbeat = 3600 min/max = NaN/NaN step = 5 4 archives(s) = [ LAST xff = 0.5 steps = 1 rows = 720 LAST xff = 0.5 steps = 12 rows = 10080 LAST xff = 0.5 steps = 180 rows = 35040 LAST xff = 0.5 steps = 2880 rows = 21900] 0 items(s) = []' for item 'MagicDimmer'.
22:11:45.372 [safeCall-1] TRACE o.o.p.r.i.RRD4jPersistenceService:331 - Using rrd definition 'default_other = GAUGE heartbeat = 3600 min/max = NaN/NaN step = 5 4 archives(s) = [ LAST xff = 0.5 steps = 1 rows = 720 LAST xff = 0.5 steps = 12 rows = 10080 LAST xff = 0.5 steps = 180 rows = 35040 LAST xff = 0.5 steps = 2880 rows = 21900] 0 items(s) = []' for item 'DemoContact'.
22:11:45.376 [safeCall-1] TRACE o.o.p.r.i.RRD4jPersistenceService:331 - Using rrd definition 'default_numeric = GAUGE heartbeat = 600 min/max = NaN/NaN step = 10 5 archives(s) = [ LAST xff = 0.5 steps = 1 rows = 360 LAST xff = 0.5 steps = 6 rows = 10080 LAST xff = 0.5 steps = 90 rows = 36500 LAST xff = 0.5 steps = 360 rows = 43800 LAST xff = 0.5 steps = 8640 rows = 3650] 0 items(s) = []' for item 'DemoNumberX'.
22:11:45.386 [safeCall-1] TRACE o.o.p.r.i.RRD4jPersistenceService:331 - Using rrd definition 'default_other = GAUGE heartbeat = 3600 min/max = NaN/NaN step = 5 4 archives(s) = [ LAST xff = 0.5 steps = 1 rows = 720 LAST xff = 0.5 steps = 12 rows = 10080 LAST xff = 0.5 steps = 180 rows = 35040 LAST xff = 0.5 steps = 2880 rows = 21900] 0 items(s) = []' for item 'DemoRollershutter'.
22:11:45.391 [safeCall-1] TRACE o.o.p.r.i.RRD4jPersistenceService:331 - Using rrd definition 'ctr24h = COUNTER heartbeat = 900 min/max = 0.0/NaN step = 60 2 archives(s) = [ AVERAGE xff = 0.5 steps = 1 rows = 480 AVERAGE xff = 0.5 steps = 10 rows = 144] 1 items(s) = [DemoNumber ]' for item 'DemoNumber'.
```

I'd hence say that this issue should be solved by that.

Note: Although I have also slightly updated the README file, please be assured that any help on further improving it would be much appreciated - so don't lean back, please ;-).

**Comment by JueBag** [comment]
Sorry for spoiling your evening and again a big THANK YOU!
Those last log-lines I haven't seen so far, and they do look promissing!
Will look into those changes tomorrow and will DEFINITLY continue with the readme!

**Comment by kaikreuzer** [comment]
> Sorry for spoiling your evening and again a big THANK YOU!

Haha, that wasn't you, but the years-old buggy code that I neglected to maintain ;-)

> What am I missing?

Yes, debugging it in the IDE isn't easy. I cannot really tell what's wrong in your setup with that error message.
Did the bundle resolution work? Did you save the bndrun file after that before launching the debug instance?
Note that even if you do everything correctly, it will complain about missing org.rrd4j package when resolving the configuration. This is because the Maven build actually embeds the libs into the bundle, but this embedding is not done by the IDE. I therefore did the build of the persistence service on the command line through Maven and then launched the debugging in the IDE - not very convenient :-/.

**Comment by ghys** [comment]
> I have to admit that I didn't yet use custom definitions with OH3 and instead tried to do sensible defaults that should suit most situations.

A quick question about that: the default persistence strategy is different when the item is a quantity type than when it's a regular type without a dimension. I had occurrences of items mistakenly defined without a dimension, and updated afterwards, but perhaps as expected the strategy was not updated; is there something that could be done about that? (apart from some documentation about how you should delete your .rrd files to start fresh with you change the type).

**Comment by kaikreuzer** [comment]
@ghys Please do not confuse persistence strategies (defined in the persist-file) with rrd datasource definitions (defined in the rrd4j.cfg file). But you are right: One a datasource files has been created for an item, this is used and won't be changed anymore by the service. This indeed will cause issues when you change the type of an existing item. But I guess that's the same for most other persistence services - if you use a specific table column in an SQL database, you'll also be in trouble.

With rrd4j being the "beginner-friendly default" service, I understand your concerns, though. But I do not have a good suggestion how to deal with that. Automatically deleting the datasource file and creating a new one is no option as valuable history might be lost. We could create a backup, which the user could manually restore if required, but that does not sound very user-friendly either. An automatic datasource migration while keeping the history in place is imho technically almost impossible :-/ .

**Comment by JueBag** [comment]
I was going to state the same and would also vote against an automatic deletion of the existing .rrd file.

Regarding the changed readme, I do note that the statement  about the required everyMinute strategy was kept, my present (so far uncomplete) tests look like the use of datasource defined step size kept by 60 seconds or less would be sufficient.

**Comment by kaikreuzer** [comment]
@JueBag Yes, I think you are right. I kept that statement not because it is correct, but because I didn't exactly know what would be a correct statement. So as I said: Any further work&improvement on the readme is very welcome!

**Comment by JueBag** [comment]
Working on it...

**Comment by JueBag** [comment]
Tested the new rrd4j persistence.
A custom datasource is now being used, however I did find a "questionable" behaviour:
When using a step size of 60' with a strategy every2Minutes I would expect rrd4j to store a value every 60 seconds in archive one, however only every 120 a value is saved.
When using a step size of 10' with a strategy every2Minutes it works as expected! 

I'm presently not sure about that.

**Comment by kaikreuzer** [comment]
> When using a sample intervall of 60 sec with a strategy every2Minutes I would expect rrd4j to store a value every 60 sec in archive one, however only every 120 sec a value is saved.

I'd say this is fully correct. If your strategy is to persist every 2 minutes, it stores a value every 2 minutes.
What you call "sample interval" is the granularity of the archive - it just means that if you persist more often than that, only one value will be in the store in the end.

**Comment by JueBag** [comment]
Thanks for answering on this closed issue.  I'll use the findings for the documetation.

I'm using the term "sample size" because it is the used term in the documentation.

> If your strategy is to persist every 2 minutes, it stores a value every 2 minutes.
Correct a NEW value is stored every minute. However according the archive setting a sample is expected every minute (with the unchanged value for every odd minute). I'm not able to explain such behaviour in conjunction with the settings for archive 1 ( `steps = 1 rows = 120`  would 120 * 60 seconds or 120*120 seconds ???)
When using one of the default databases it is working exactly that way, a (possibly unchanged) sample is taken every 10 (or 5) seconds and an updated value is stored by openHAB according the whatsoever strategy.
Using 10 or 20 seconds in a custom Database setup is working as expected as well, starting with 30 seconds I see the above mentioned behaviour.

My personell opinion on that:

Explaining this different behaviour to users is (at least) error prone! 
I'd rather suggest the ADVICE the use of sample sizes 5,10 or20 seconds (instead of the ADVICE to use an everyMinute Strategy).

**Comment by kaikreuzer** [comment]
@JueBag I saw that you had added the following text as a breaking change in the wiki:
```
The updated rrd4j persistence service got changed **default** database setups. Users of the **default** database setup of rrd4j Version 1 should note the changed setup for the now three different defaults. By the further use of rrd4j with no custom database setup the history would be lost!
In order to keep the history a custom setup, using the former default settings, is needed.
```
I removed it for the moment, because I honestly didn't understand what it wanted to tell the user.
I assume it is somehow related to this issue here?
In general I would hope that no history is lost/overwritten, when users do an upgrade. Do we have any issue there?

**Comment by JueBag** [comment]
In **my** understanding the continued use of .rrd files which have been created with the "old" version and its default database setup in OH3 would require an explicit use of  a custom database setup which resembles the version 1 default setup.
Or am I completely on the wrong track?

**Comment by JueBag** [comment]
@kaikreuzer 
I have to admit that I proved myself WRONG on that!
I was in the wrong believe that rrd4j would only be able to store data in a specific custom way when the defining rrd4j.cfg would exist.
I finally managed to the default setup of OH3 back and tested the above case.  Dropping in an old .rrd file (created in OH2 with the default setup) results in:

- the old file gets continued as setup (unchanged file-size,...)
- all other .rrd files (automatically created ones by the default strategy) are getting updated as well.
